A2017/tensorflow.ipynb

```diff
@@ -0,0 +1,1357 @@
+    "\n",
+    "We instantiate the `tf.train.GradientOptimizer` by passing it a scalar learning rate. Note that the learning rate itself can be symbolic, and is allowed to vary across a session."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 26,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can then call the optimizer's `minimize` function to obtain an op with, when evaluated, does a gradient descent step on the variables we specified:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "update_op = optimizer.minimize(loss=z, var_list=tf.trainable_variables())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here we took advantage of the fact that all variables created via `tf.get_variable` can be accessed as a list using the `tf.trainable_variables` function.\n",
+    "\n",
+    "(Note: you can pass `trainable=False` to `tf.get_variable` to exclude a certain variable from ending up in that list.)\n",
+    "\n",
+    "The code then proceeds as before:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 28,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "x = 0.52, y = -1.75, z = 7.49\n",
+      "x = 1.11, y = -2.25, z = 2.70\n",
+      "x = 1.47, y = -2.55, z = 0.97\n",
+      "x = 1.68, y = -2.73, z = 0.35\n",
+      "x = 1.81, y = -2.84, z = 0.13\n",
+      "x = 1.89, y = -2.90, z = 0.05\n",
+      "x = 1.93, y = -2.94, z = 0.02\n",
+      "x = 1.96, y = -2.97, z = 0.01\n",
+      "x = 1.98, y = -2.98, z = 0.00\n",
+      "x = 1.99, y = -2.99, z = 0.00\n"
+     ]
+    }
+   ],
+   "source": [
+    "with tf.Session() as session:\n",
+    "    session.run(tf.global_variables_initializer())\n",
+    "\n",
+    "    for _ in range(10):\n",
+    "        session.run(update_op)\n",
+    "        x_val, y_val, z_val = session.run([x, y, z])\n",
+    "        print('x = {:4.2f}, y = {:4.2f}, z = {:4.2f}'.format(x_val, y_val, z_val))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Control flow\n",
+    "\n",
+    "TensorFlow implements symbolic operations for executing common control flow structures, like if-statements and while loops.\n",
+    "\n",
+    "This is usually where people get confused: despite appearances, there is a *big* difference between a regular control flow statement and its symbolic counterpart. **When working with control flow ops, you should always keep in mind that the code you write _defines the computation graph_, it does not _execute_ that computation.**\n",
```

Comment:
A way I have heard that described is that it is _metaprogramming_: we are using one language (Python) to build a program in a different language (TF's dataflow). It's not unlike the C preprocessor macros in that sense: there is a big difference between `#if ...` and `if(...)`.

Reply:
That's a good analogy as well!